crates/warg-client/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+use semver::Version;
+use url::Url;
+
+/// A client for interacting with a registry
+#[allow(dead_code)]
+pub struct Client<RegistryCache, ValidationCache> {
+    registry: DomainName,
+    known_roots: Vec<Root>,
+    registry_cache: RegistryCache,
+    validation_cache: ValidationCache,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DomainName(String);
+
+#[allow(dead_code)]
+pub struct Root {
+    bytes: Vec<u8>
+}
+
+pub trait Cache {
+    /// Insert a value into the clients cache.
+    fn insert(&mut self, key: Vec<u8>, value: Vec<u8>);
+
+    /// Lookup a previously inserted cache value.
+    fn lookup(&self, key: &[u8]) -> Option<Vec<u8>>;
+}
+
+impl<RegistryCache, ValidationCache> Client<RegistryCache, ValidationCache>
+where
+    RegistryCache: Cache,
+    ValidationCache: Cache
+{
+    /// Create a new client with no previous state
+    pub fn new(registry: DomainName, registry_cache: RegistryCache, validation_cache: ValidationCache) -> Self {
+        Self {
+            registry,
+            known_roots: Vec::new(),
+            registry_cache,
+            validation_cache
+        }
+    }
+
+    /// Check what registry is being used
+    pub fn registry(&self) -> DomainName {
+        self.registry.clone()
+    }
+
+    /// Serialize the client state
+    pub fn to_state(&self) -> Vec<u8> {
+        unimplemented!()
+    }
+
+    /// Construct a registry client with prior state
+    pub fn from_state(registry: DomainName, state: Vec<u8>, registry_cache: RegistryCache, validation_cache: ValidationCache) -> Self {
```

Comment:
Seems like you might already be getting into enough complexity here that a `ClientBuilder` might make sense.